### ISSUE TITLE
Quake Mode should remember its size

### ISSUE THREAD

**Issue by gamagan** [microsoft/terminal#12594]
# Description of the new feature/enhancement

I often want the Quake Mode terminal window to be different sizes,  depending on what I'm doing at the time. However, the window size always resets upon closing and reopening.

There can be a setting in the config file to allow the Quake Mode window to remember its size upon opening/closing.

**Comment by zadjii-msft** [comment]
Huh, coulda swore I had a dupe for this. 

I don't think we should even have a setting for this, it should probably just always do this.

**Comment by gamagan** [comment]
I have a 3 monitor setup. I hadn't notice this before, but the window size is remembered on 1 of the monitors (not the primary one). It's remembered as long as you toggle the window only on that monitor. Once you open it in a different monitor, then the size resets in the remembered one, too.

**Comment by nakdeyes** [comment]
Looking forward to a fix for this one day! Quake mode in windows terminal made me finally bail on CMDR / Cygwin combo I have been using for years. Just need this freature to match the functionality there.

**Comment by nakdeyes** [comment]
Hot tip until this feature comes online: don’t forget about F11 to full screen while in quake mode.

it nicely fills the screen and it does appear that quake mode will remember this setting and can hide/reveal back into full screen.

helps me a lot anyways as my main use case is “I want it taller”. I at least have 2 sizes to choose between now!

**Comment by riotrah** [comment]
![Screenshot_20221207-213159.png](https://user-images.githubusercontent.com/22646419/206364690-aa32ed64-4e76-45d4-99b4-4f9a15cad1cc.png)

Nooooooooo! 😭

**Comment by riotrah** [comment]
> Hot tip until this feature comes online: don’t forget about F11 to full screen while in quake mode.
> 
> it nicely fills the screen and it does appear that quake mode will remember this setting and can hide/reveal back into full screen.
> 
> helps me a lot anyways as my main use case is “I want it taller”. I at least have 2 sizes to choose between now!

Brilliant, gonna do that now

**Comment by rationalthinker1** [comment]
> Hot tip until this feature comes online: don’t forget about F11 to full screen while in quake mode.
> 
> it nicely fills the screen and it does appear that quake mode will remember this setting and can hide/reveal back into full screen.
> 
> helps me a lot anyways as my main use case is “I want it taller”. I at least have 2 sizes to choose between now!

Unforunately, it doesn't work on multiple monitors.

**Comment by gbiagini** [comment]
Just wanted to +1, this feature would make quake mode complete. 

Even allowing the full screen to persist on multi-monitor setups would be good enough.

**Comment by Keropon** [comment]
Would love this feature to be honest, since I discovered the Quake mode I've been using it daily, so this would only just make it better

**Comment by GreatestFool** [comment]
I'd really appreciate this and also just being able to set a default size in the config. I generally decrease the Quake Mode height to about 25% of my screen instead of half.

**Comment by shayded-exe** [comment]
How is this still an issue in 2024? Makes quake mode unusable on laptops.

**Comment by slrgt** [comment]
any update on this?

**Comment by zadjii-msft** [comment]
Nope. We'll make sure to update this thread when there is. In the meantime, might I recommend the Subscribe button?
![image](https://user-images.githubusercontent.com/18356694/91237459-5cbb0c80-e700-11ea-9347-b9b1ec2813b1.png)
That way you'll be notified of any updates to this thread, without needlessly pinging everyone on this thread ☺️

Alternatively, if anyone is particularly passionate about this, I could help point to the right code in `IslandWindow.cpp` to start investigating.

**Comment by CodeVogel** [comment]
> Alternatively, if anyone is particularly passionate about this, I could help point to the right code in `IslandWindow.cpp` to start investigating.

Yes please! I have no prior experience in this codebase, but this would make it perfect for me when used in conjunction with TMUX.
It would offer me the option to quickly collapse and restore all my development windows and summon them back again. Having a dedicated button shortcut to summon my terminal to my current window would be much better than leaving a terminal window in fullscreen and having to alt-tab back to it.

**Comment by zadjii-msft** [comment]
@CodeVogel for what it's worth, you may be more interested in just plain-old [`globalSummmon`](https://learn.microsoft.com/en-us/windows/terminal/customize-settings/actions#global-summon), which you can use with any window (not just a quake one):

I have it bound like: 
```json
        {
            "command":
            {
                "action": "globalSummon",
                "monitor": "any"
            },
            "keys": "win+sc(41)"
        },
```

but sounds like you want `"monitor": "toCurrent"`

**Comment by CodeVogel** [comment]
@zadjii-msft awesome! that is pretty much exactly what I was looking for.

The `toCurrent` strategy behaves a little weird, as when I dismiss the terminal on my right monitor and have nothing else in focus on the right monitor, it opens on the left monitor, but that's unrelated to this issue.

**Comment by mmaazahmed** [comment]
# My solution:
step 1) add the following setting to your settings json:
  {
            "command":
            {
                "action": "globalSummon",
                "monitor": "any"
            },
            "keys": "win+sc(41)"
        },

step 2: go to settings>appearance > enable `Hide terminal in the notification area when minimized` and `automatically hide window`

step 3: go to start up > launch parameters > set launch mode to focus.

you now have a `quake mode` that works with windows tiling and rememebers its size and position. Here's a short demo:

![Image](https://github.com/user-attachments/assets/e0edc2dc-0505-44dd-8406-03dad1a6a1d6)
https://github.com/user-attachments/assets/eeb974e4-c149-42bc-b6ec-283e3882269b

**Comment by romanbeldent** [comment]
Hey, no fix for quake height size or for quake remembering its size when re-opening ?

**Comment by zadjii-msft** [comment]
We'd be happy to accept a PR for this, and if anyone's willing to contribute, I could try and point them at where in the code I think this might be!

**Comment by metal-gabe** [comment]
> We'd be happy to accept a PR for this, and if anyone's willing to contribute, I could try and point them at where in the code I think this might be!

I'd be willing to give it a shot 🙂 a point in the right code direction would be greatly appreciated

**Comment by zadjii-msft** [comment]
I'd probably start looking near this function:

https://github.com/microsoft/terminal/blob/fb668f3a4ad14a876585c2f1603880dd159f64eb/src/cascadia/WindowsTerminal/IslandWindow.cpp#L1678-L1687

That's what calculates the size the quake window should be at. But instead of it just always being half, we'd need to somehow know
* That someone resized it to be a different size
* what percent of the monitor that is now (so that on different monitors, we'll still use the same relative height)
* and then the next time we're summoned, use that percent (instead of always 50%)

**Comment by foreignmeloman** [comment]
Maybe it's worth to have a setting to just define the precentage? I don't see a scenario where someone would manually adjust the quake termial size constantly.

**Comment by FxAlex** [comment]
> Maybe it's worth to have a setting to just define the precentage? I don't see a scenario where someone would manually adjust the quake termial size constantly.

I disagree. It would certainly be better to save the height each time it changes.

**Comment by Keropon** [comment]
It's a terminal, that might or might not be okay with 50% of the total screen state. **I'd personally prefer to set a size in pixels**, since I'd know more or less when I have 10 or 20 lines to work with, just by multiplying the font size.

What I would really like tho is that the terminal size could be saved by a screen-basis, so if I summon the terminal while in my 4k or my 1080p screen it will keep the size I set the last time for each.

**Comment by zadjii-msft** [comment]
My pro-tip for 3p contributions is _start small_.

* _Start_ by just making a change that persists the relative size when the user changes it. 
* Then maybe a follow-up PR could make that persist per-monitor. 
* Then another PR could add a setting for the "default" size to open at.
* then maybe another PR takes that setting and parses any of { `50%`, `400dip`, `600px`, `32` (cells) } and does The Right Thing depending on the unit

But always start small and make incremental improvement. I wouldn't want meaningful piecewise improvements to be locked behind attempting to boil the ocean

**Comment by metal-gabe** [comment]
> My pro-tip for 3p contributions is _start small_.
> 
> * _Start_ by just making a change that persists the relative size when the user changes it.
> * Then maybe a follow-up PR could make that persist per-monitor.
> * Then another PR could add a setting for the "default" size to open at.
> * then maybe another PR takes that setting and parses any of { `50%`, `400dip`, `600px`, `32` (cells) } and does The Right Thing depending on the unit
> 
> But always start small and make incremental improvement. I wouldn't want meaningful piecewise improvements to be locked behind attempting to boil the ocean

will do, thanks for the advice!
